File: 13-02-2026/problem_4_valid_parentheses/correct/valid_parentheses.java
// PROBLEM 1: LONGEST VALID PARENTHESES - CORRECT JAVA
// ============================================
import java.util.*;

class Solution {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        if (!sc.hasNext()) {
            System.out.println(0);
            return;
        }
        
        String s = sc.next();
        Stack<Integer> st = new Stack<>();
        st.push(-1);
        int best = 0;
        
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                st.push(i);
            } else {
                st.pop();
                if (!st.isEmpty()) {
                    best = Math.max(best, i - st.peek());
                } else {
                    st.push(i);
                }
            }
        }
        
        System.out.println(best);
    }
}


// =============================================

File: 13-02-2026/problem_4_valid_parentheses/correct/valid_parentheses.py
# PROBLEM 1: LONGEST VALID PARENTHESES - CORRECT PYTHON
# ============================================
import sys

s = sys.stdin.readline().strip()

if not s:
    print(0)
else:
    stack = [-1]
    best = 0
    
    for i in range(len(s)):
        if s[i] == '(':
            stack.append(i)
        else:
            stack.pop()
            if stack:
                best = max(best, i - stack[-1])
            else:
                stack.append(i)
    
    print(best)


# =============================================